models/rnnt.py

```diff
@@ -0,0 +1,206 @@
+      added = 0
+      while not_blank and added < 30:
+        label = Tensor([[labels[-1] if labels[-1] <= 28 else labels[-1] - 1]]) if len(labels) > 0 else None
+        g, h_, c_ = self.prediction(label, h, c)
+        j = self.joint(logit, g)[:, 0, 0, :][0, :].numpy()
+
+        k = np.argmax(j, axis=0)
+        not_blank = k != 28
+        if not_blank:
+          labels.append(k)
+          h, c = h_, c_
+        added += 1
+    return labels
+
+  def load_from_pretrained(self):
+    fn = Path(__file__).parent.parent / "weights/rnnt.pt"
+    download_file("https://zenodo.org/record/3662521/files/DistributedDataParallel_1576581068.9962234-epoch-100.pt?download=1", fn)
+
+    import torch
+    state_dict = torch.load(open(fn, "rb"), map_location="cpu")["state_dict"]
+
+    # encoder
+    for i in range(2):
+      self.encoder.pre_rnn.cells[i].weights_ih.assign(state_dict[f"encoder.pre_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].weights_hh.assign(state_dict[f"encoder.pre_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].bias_ih.assign(state_dict[f"encoder.pre_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].bias_hh.assign(state_dict[f"encoder.pre_rnn.lstm.bias_hh_l{i}"].numpy())
+    for i in range(3):
+      self.encoder.post_rnn.cells[i].weights_ih.assign(state_dict[f"encoder.post_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].weights_hh.assign(state_dict[f"encoder.post_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].bias_ih.assign(state_dict[f"encoder.post_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].bias_hh.assign(state_dict[f"encoder.post_rnn.lstm.bias_hh_l{i}"].numpy())
+
+    # prediction
+    self.prediction.emb.weight.assign(state_dict["prediction.embed.weight"].numpy())
+    for i in range(2):
+      self.prediction.rnn.cells[i].weights_ih.assign(state_dict[f"prediction.dec_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.prediction.rnn.cells[i].weights_hh.assign(state_dict[f"prediction.dec_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.prediction.rnn.cells[i].bias_ih.assign(state_dict[f"prediction.dec_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.prediction.rnn.cells[i].bias_hh.assign(state_dict[f"prediction.dec_rnn.lstm.bias_hh_l{i}"].numpy())
+
+    # joint
+    self.joint.l1.weight.assign(state_dict["joint_net.0.weight"].numpy())
+    self.joint.l1.bias.assign(state_dict["joint_net.0.bias"].numpy())
+    self.joint.l2.weight.assign(state_dict["joint_net.3.weight"].numpy())
+    self.joint.l2.bias.assign(state_dict["joint_net.3.bias"].numpy())
+
+
+class LSTMCell:
+  def __init__(self, input_size, hidden_size):
+    self.hidden_size = hidden_size
+
+    self.weights_ih = Tensor.uniform(hidden_size * 4, input_size)
+    self.bias_ih = Tensor.uniform(hidden_size * 4)
+    self.weights_hh = Tensor.uniform(hidden_size * 4, hidden_size)
+    self.bias_hh = Tensor.uniform(hidden_size * 4)
```

Comment:
(Not a review just a question) Why use tensors instead of nn.Linear here?

Reply:
maybe its cleaner to use linear here but the weight initialization for linear is different than the weight initialization for LSTM at least in pytorch.